### DISCUSSION TITLE
GPIO line status from Basler camera

### DISCUSSION THREAD

**Discussion by cvargas4** [bonsai-rx/bonsai#1871] [Q&A] [answered]
Hello!
I'm trying to set up a recording system with centralized triggers from an Arduino going to several devices, including some Basler cameras. Is there a way to get the GPIO input status from the Basler cameras using the Pylon package in Bonsai? I see a similar question was asked before (https://groups.google.com/g/bonsai-users/c/6Xb2hJ_GYak/m/j2GCNG8xFQAJ), but I can't figure out how exactly to use the GrabResult.ChunkData to recover LineStatusAll.

Thank you

**Comment by chiyu1203** [comment]
Hi I used Arduino to trigger multiple basler camera together on Bonsai. But I dont know what you meant by "GrabResult.ChunkData to recover LineStatusAll". Could you describe your problem more clearly? For example, what you aim to achieve in the end.

**Reply by cvargas4** [reply]
Hi! Sure, what I want to do is to read in and save the trigger state for each frame, so that later/offline I can align the videos with other data streams receiving the same pulses from the Arduino. Is this how you use your system?

Regarding the ChunkData, what I meant by the GrabResult.Chunkdata and LineStatusAll is in reference to this documentation from Basler ([https://docs.baslerweb.com/data-chunks](https://docs.baslerweb.com/data-chunks)). It seems like there's a way to parse the parameter ChunkData and identify the trigger state of a GPIO line? In Bonsai when right clicking PylonCapture there's an option for GrabResult.ChunkData, but GrabResult.ChunkData outputs a string.  I assume means it can't be manipulated further in Bonsai? In any case, I was thinking this might be a way to get the GPIO state of a Basler camera, but I don't know if this is feasible.

![image](https://github.com/bonsai-rx/bonsai/assets/33707956/6d049358-0b04-4712-a1ab-adaf3e51cabc)

**Reply by chiyu1203** [reply, marked as answer]
Firstly, I have never used ChunkData so dont know about that. Instead, I use GrabResult.Timestamp to get the timestamp when a data point reached Bonsai.
I was a bit confused by your original intention because I remember that you can only set Basler cameras to either send out pulse when it acquire a frame or trigger a frame acquisitions when it receive pulse (with the specified GPIO lines and its respective ground line). This is done on Pylong viewer and you need to save your setting into a feature file so that it can be used in Pylon Capture on Bonsai. Whether you want to use arduino to trigger your camera, meanwhile sending a pulse to your ephys acquisition system, or asking the Basler camera to send pulse to the ephys, it's up to you. I use arduino but "only to trigger" camera to acquire new frames at the same time. When aligning video data with ephys data, I use [Synchronizing video with ephys using an LED](https://bonsai-rx.org/neuroaspects-2017/worksheets/synching/) for stimulus-evoked responses and use [barcode](https://optogeneticsandneuralengineeringcore.gitlab.io/ONECoreSite/projects/DAQSyncro/BarcodeGeneration/) for others. I can be wrong but the idea of not to directly logging the state of camera to ephys data stream is simply because with unique pattern of barcodes (and usually people film behaviours at lower frame rate), it is much easier to align two different data stream.
Below is an example of how I use arduino to trigger Basler camera on Bonsai (I coded the arduino code to receive strings from Bonsai so that it knows when exactly to start and stop)
[multple_camera_tracking.bonsai.zip](https://github.com/user-attachments/files/16090640/multple_camera_tracking.bonsai.zip)
